### PR TITLE
While resolving assemblies, populate the RequestingAssembly property of ResolveEventArgs

### PR DESCRIPTION
**Pull request by Tragetaschen**
Currently, the RequestingAssembly property of ResolveEventArgs [is not populated](https://github.com/mono/mono/blob/master/mcs/class/corlib/System/AppDomain.cs#L1237) when resolving assemblies.
This PR adds the necessary plumbing to transport the requesting assembly through the hooks.

This fixes [a problem in ASP.NET 5](https://github.com/aspnet/XRE/issues/1048) and fixes the behavioral differences as observed [here](https://github.com/davidfowl/Mono-Assembly-Resolve-Issue)

### CODE REVIEW

**Comment by monoadmin** [comment]
Hello! I'm the build bot for the Mono project. I need approval from a Mono team member to build this pull request. A team member should reply with "approve" to approve a build of this pull request, "whitelist" to whitelist this and all future pull requests from this contributor, or "build" to explicitly request a build, even if one has already been done. Contributors can ignore this message.

File: mono/metadata/assembly.c [hunk 1]
```diff
@@ -1764,7 +1764,7 @@ mono_assembly_load_from_full (MonoImage *image, const char*fname,
 	 * assemblies lock.
 	 */
 	if (ass->aname.name) {
-		ass2 = mono_assembly_invoke_search_hook_internal (&ass->aname, refonly, FALSE);
+		ass2 = mono_assembly_invoke_search_hook_internal (&ass->aname, NULL, refonly, FALSE);
```
Comment (alexrp) [inline mono/metadata/assembly.c]: This (and the hunk below) doesn't seem right. `mono_assembly_load_from_full` is invoked from e.g. `ves_icall_System_AppDomain_LoadAssemblyRaw` which is used in `AppDomain.Load ()`. It's also used in many other API functions in `assembly.c` which should probably also have a `requesting` parameter added.

File: mono/metadata/appdomain.c [hunk 2]
```diff
@@ -1983,7 +1984,7 @@ ves_icall_System_AppDomain_LoadAssembly (MonoAppDomain *ad,  MonoString *assRef,
 	if (!parsed) {
 		/* This is a parse error... */
 		if (!refOnly)
-			refass = mono_try_assembly_resolve (domain, assRef, refOnly);
+			refass = mono_try_assembly_resolve (domain, assRef, NULL, refOnly);
```
Comment (alexrp) [inline mono/metadata/appdomain.c]: These two hunks also seem odd to me.

**Comment by alexrp** [comment]
approve

**Comment by marek-safar** [comment]
Please write a test for this feature

File: mono/metadata/assembly.c (hunk 1 above)
Comment (Tragetaschen) [inline mono/metadata/assembly.c]: See the test for .NET-framework's behavior. The event handler is called with `RequestingAssembly == null` for assemblies loaded with `AppDomain.Load` (or any of the other possible entry points). I'd guess the reasoning behind this is that explicitly loading does not mean requesting by an assembly, but requesting by user code.

File: mono/tests/assemblyresolve_event4.cs [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+			Console.WriteLine ("Caught exception: {0}", ex);
+			return 1;
+		}
+
+		return 0;
+	}
+
+	static Assembly MyResolveEventHandler (object sender, ResolveEventArgs args)
+	{
+		var path = String.Format ("{0}{1}assemblyresolve{1}deps{1}", Directory.GetCurrentDirectory (), Path.DirectorySeparatorChar);
```
Comment (akoeplinger) [inline mono/tests/assemblyresolve_event4.cs]: Use Path.Combine() instead of the string format.

Comment (Tragetaschen) [inline mono/tests/assemblyresolve_event4.cs]: Yeah, I started with the *event3.cs and ported this line

**Comment by kumpera** [comment]
Hi,

There are a couple of issues with this patch.

First, it must be under the MIT license for us to merge it.
Second, it does change the public API and this will break embedders.

We can't merge unless the two are addressed.

File: mono/metadata/assembly.h [hunk 4]
```diff
@@ -13,9 +13,11 @@ MONO_API MonoAssembly *mono_assembly_open_full (const char *filename,
 				       	MonoImageOpenStatus *status,
 					mono_bool refonly);
 MONO_API MonoAssembly* mono_assembly_load       (MonoAssemblyName *aname, 
+                                       	MonoAssembly *requesting,
                                        	const char       *basedir, 
 				     	MonoImageOpenStatus *status);
 MONO_API MonoAssembly* mono_assembly_load_full (MonoAssemblyName *aname, 
+                                       	MonoAssembly *requesting,
                                        	const char       *basedir, 
 				     	MonoImageOpenStatus *status,
```
Comment (migueldeicaza) [inline mono/metadata/assembly.h]: Both of those changes break the embedding API.

This requires new API entry points to be added to provide the missing information.

**Comment by Tragetaschen** [comment]
Needless to say I didn't think about that.

However now, the default postload hook is invoked manually, which makes the change a bit more involved.

This contribution is made under the terms of the MIT license.

**Comment by kumpera** [comment]
I think this PR is a solid foundation for fixing this issue. There's one last requirement related to callback ordering that I'll fix myself and push together with this PR.

Thanks for the contribution.

**Comment by Tragetaschen** [comment]
Because I made this one specific ordering decision explicitly:

The default postload search is executed last on purpose: Newly registered hooks are pushed to the front of the linked list. This means when they are registered after the appdomain is initialized, they are executed before any of the default hooks.

**Comment by kumpera** [comment]
I merged your previous patch plus a fix to retain the original ordering.

**Comment by Tragetaschen** [comment]
And I get just how subtle these ordering issues can be. Yikes!

Btw: Wheeee :joy:

**Pull request closed without merging**